slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -22,6 +22,37 @@ ::grpc::Status SLAMServiceImpl::GetPosition(ServerContext *context,
     LOG(ERROR) << "GetPosition is not yet implemented.\n";
     return grpc::Status(grpc::StatusCode::UNIMPLEMENTED,
                         "GetPosition is not yet implemented.");
+    cartographer::transform::Rigid3d global_pose;
+    {
+        std::lock_guard<std::mutex> lk(map_builder_mutex);
```

Comment:
Actually this should not be locking the `map_builder_mutex`. We do instead probably need a `slam_mutex` that'll lock when we change/get the global pose.

Reply:
Agreed. Also, would it be better to store the trajectory nodes as a member variable and update them when we update `latest_global_pose`?  This would ensure that `GetMap` doesn't need to block if cartographer is currently processing data. Should we file a ticket for this? I wish I had thought of this during the review for https://viam.atlassian.net/browse/DATA-147 or https://viam.atlassian.net/browse/DATA-159. This may also be part of why it takes a long time to paint a jpeg map (https://viam.atlassian.net/browse/DATA-661).

Reply:
I will file a ticket for this.

Reply:
https://viam.atlassian.net/browse/DATA-706